webapp/spdx.py

```diff
@@ -0,0 +1,434 @@
+    ('NCSA', 'University of Illinois/NCSA Open Source License'),
+    ('Vim', 'Vim License'),
+    ('VOSTROM', 'VOSTROM Public License for Open Source'),
+    ('VSL-1.0', 'Vovida Software License v1.0'),
+    ('W3C-20150513',
+     'W3C Software Notice and Document License (2015-05-13)'),
+    ('W3C-19980720', 'W3C Software Notice and License (1998-07-20)'),
+    ('W3C', 'W3C Software Notice and License (2002-12-31)'),
+    ('Wsuipa', 'Wsuipa License'),
+    ('Xnet', 'X.Net License'),
+    ('X11', 'X11 License'),
+    ('Xerox', 'Xerox License'),
+    ('XFree86-1.1', 'XFree86 License 1.1'),
+    ('xinetd', 'xinetd License'),
+    ('xpp', 'XPP License'),
+    ('XSkat', 'XSkat License'),
+    ('YPL-1.0', 'Yahoo! Public License v1.0'),
+    ('YPL-1.1', 'Yahoo! Public License v1.1'),
+    ('Zed', 'Zed License'),
+    ('Zend-2.0', 'Zend License v2.0'),
+    ('Zimbra-1.3', 'Zimbra Public License v1.3'),
+    ('Zimbra-1.4', 'Zimbra Public License v1.4'),
+    ('Zlib', 'zlib License'),
+    ('zlib-acknowledgement', 'zlib/libpng License with Acknowledgement'),
+    ('ZPL-1.1', 'Zope Public License 1.1'),
+    ('ZPL-2.0', 'Zope Public License 2.0'),
+    ('ZPL-2.1', 'Zope Public License 2.1'),
+]
+
+LICENSE_EXCEPTIONS = [
+    ('Autoconf-exception-2.0', 'Autoconf exception 2.0'),
+    ('Autoconf-exception-3.0', 'Autoconf exception 3.0'),
+    ('Bison-exception-2.2', 'Bison exception 2.2'),
+    ('Classpath-exception-2.0', 'Classpath exception 2.0'),
+    ('eCos-exception-2.0', 'eCos exception 2.0'),
+    ('Font-exception-2.0', 'Font exception 2.0'),
+    ('GCC-exception-2.0', 'GCC Runtime Library exception 2.0'),
+    ('GCC-exception-3.1', 'GCC Runtime Library exception 3.1'),
+    ('WxWindows-exception-3.1', 'WxWindows Library Exception 3.1'),
+]
+
+LICENSE_PROPRIETARY = 'Proprietary'
+LICENSE_OTHER = 'Other Open Source'
+
+# store custom licenses (non SPDX)
+CUSTOM_LICENSES = [
+    (LICENSE_PROPRIETARY, 'Proprietary'),
+    (LICENSE_OTHER, 'Other Open Source'),
+    ('AGPL-3.0+', 'GNU Affero General Public License v3.0 or later'),
+    ('GPL-2.0+', 'GNU General Public License v2.0 or later'),
+    ('GPL-3.0+', 'GNU General Public License v3.0 or later'),
+    ('LGPL-2.1+', 'GNU Lesser General Public License v2.1 or later'),
+    ('LGPL-3.0+', 'GNU Lesser General Public License v3.0 or later'),
+]
+
+VALID_LICENSES = dict(LICENSES + CUSTOM_LICENSES)
+
+
+def _get_parser():
+    symbols = ([LicenseSymbol(k) for k in VALID_LICENSES.keys()] +
+               [LicenseSymbol(k, is_exception=True)
+                for k in dict(LICENSE_EXCEPTIONS)])
+    licensing = Licensing(symbols=symbols)
+    return licensing
+
+
+def validate_spdx_license_expression(expression):
+    """Validate and simplify SPDX expression."""
+    if not expression:
+        return expression
+
+    licensing = _get_parser()
+    try:
+        parsed = licensing.parse(expression, validate=True, strict=True)
+    except ExpressionError as e:
+        msg = str(e)
```

Comment:
was: `e.message` now I just `str` the excpection since message is not valid anymore